SpaceTimeStack: Account for MPI imbalance when applying printing threshold and add test

vbrunini opened the pull request:
So that entries with a low average time but high imbalance are reported since they are often the cause of significant performance problems due to load imbalance.

Also fix missing deep_copy hook exposure.

vbrunini committed:
SpaceTimeStack: Account for MPI imbalance when applying printing threshold.

So that entries with a low average time but high imbalance are reported
since they are often the cause of significant performance problems due
to load imbalance.

vbrunini committed:
SpaceTimeStack: Account for MPI imbalance when applying printing threshold.

So that entries with a low average time but high imbalance are reported
since they are often the cause of significant performance problems due
to load imbalance.

vbrunini committed:
SpaceTimeStack: Fix format.

vbrunini committed:
SpaceTimeStack: Fix format.

vbrunini committed:
SpaceTimeStack: Fix failure when MPI not initialized.

vbrunini committed:
SpaceTimeStack: Fix failure when MPI not initialized.

vlkale commented:
> …shold.
> 
> 
> 
> So that entries with a low average time but high imbalance are reported since they are often the cause of significant performance problems due to load imbalance.
> 
> 
> 
> Also fix missing deep_copy hook exposure.

@vbrunini 

Thanks for this. I am curious if you can provide an example code with the problem here in the PR description. Even better would be a set of tests via, say, Google Test, to check that those entries with low average time but high imbalance are indeed being reported.

A side note is that the fix for the missing deep_copy hook you have should ideally be a separate PR, as it looks conceptually unrelated to the change related to MPI load imbalance?

Thanks!

vbrunini committed:
SpaceTimeStack: Add test for thresholding including imbalance.

vbrunini committed:
SpaceTimeStack: Add test for thresholding including imbalance.

vbrunini commented:
> > …shold.
> > So that entries with a low average time but high imbalance are reported since they are often the cause of significant performance problems due to load imbalance.
> > Also fix missing deep_copy hook exposure.
> 
> @vbrunini
> 
> Thanks for this. I am curious if you can provide an example code with the problem here in the PR description. Even better would be a set of tests via, say, Google Test, to check that those entries with low average time but high imbalance are indeed being reported.
> 
> A side note is that the fix for the missing deep_copy hook you have should ideally be a separate PR, as it looks conceptually unrelated to the change related to MPI load imbalance?
> 
> Thanks!

Added a test (and split off the deep copy parts into #285)

vbrunini committed:
SpaceTimeStack: Format.

vbrunini committed:
SpaceTimeStack: Format.

masterleinad commented on tests/space-time-stack/test_deep_copy.cpp.orig:
I'm pretty sure you don't want to keep this file.

vbrunini committed:
SpaceTimeStack: Fix test compile & delete extra file.

vbrunini committed:
SpaceTimeStack: Fix test compile & delete extra file.

masterleinad commented on profiling/space-time-stack/kp_space_time_stack.cpp:
```diff
@@ -336,9 +336,11 @@ struct StackNode {
                        std::string const& child_indent,
                        double tree_time) const {
-    auto percent = (total_runtime / tree_time) * 100.0;
+    const double comm_size = total_runtime / avg_runtime;
+    auto threshold_percent = ((max_runtime * comm_size) / tree_time) * 100.0;
+    auto percent           = (total_runtime / tree_time) * 100.0;
```
Can you elaborate on why this is sensible? I understand the previous code but now we essentially do
```
((max_runtime / avg_runtime) * percent  < outout_threshold
```
In other words, why does multiplying by `max_runtime / avg_runtime` do what you want? Wouldn't you want to do something like
```
if (percent > first_threshold || (high_imbalance && percent > second_threshold))
  report
```
? It's not quite clear to me that this second threshold should be the same as the orignal one divided by the imbalance. Maybe it's sensible to report all kernels that have an imbalance greater than a certain factor?

vbrunini replied:
The thinking is to identify any kernels/regions that would be above the 0.1% threshold if we only looked at the runtimes on that rank. Assuming that total_runtime is evenly distributed across ranks (which I think is basically always true), then `rank_wall_time = tree_time / comm_size` and therefore `max_runtime / rank_wall_time == (max_runtime * comm_size) / tree_time`.

masterleinad replied:
OK. You basically want to use `max_runtime` instead of `avg_runtime` for comparing with the threshold value. I think that's sensible. Not sure if we need a configuration option.
Note that there also is `print_json_recursive` that needs consistent logic.

masterleinad committed:
Merge branch 'develop' into vbrunini/ststack_imbalance_threshold

masterleinad commented on profiling/space-time-stack/kp_space_time_stack.cpp:
```diff
@@ -336,9 +336,11 @@ struct StackNode {
                        std::string const& child_indent,
                        double tree_time) const {
-    auto percent = (total_runtime / tree_time) * 100.0;
+    const double comm_size = total_runtime / avg_runtime;
```
You might as well compute this properly once and store it in a member variable.

vbrunini committed:
SpaceTimeStack: Address review comments.

vbrunini committed:
SpaceTimeStack: Address review comments.

vbrunini committed:
SpaceTimeStack: Format.

vbrunini committed:
SpaceTimeStack: Format.

vbrunini commented:
Any suggestions for what to do about the CI failure from the tests running as root?
```
--------------------------------------------------------------------------
mpirun has detected an attempt to run as root.

Running as root is *strongly* discouraged as any mistake (e.g., in
defining TMPDIR) or bug can result in catastrophic damage to the OS
file system, leaving your system in an unusable state.

We strongly suggest that you run mpirun as a non-root user.

You can override this protection by adding the --allow-run-as-root option
to the cmd line or by setting two environment variables in the following way:
the variable OMPI_ALLOW_RUN_AS_ROOT=1 to indicate the desire to override this
protection, and OMPI_ALLOW_RUN_AS_ROOT_CONFIRM=1 to confirm the choice and
add one more layer of certainty that you want to do so.
We reiterate our advice against doing so - please proceed at your own risk.
--------------------------------------------------------------------------
```

masterleinad committed:
Allow MPI to run as root

masterleinad committed:
Allow MPI to run as root

masterleinad commented on tests/CMakeLists.txt:
```diff
@@ -24,6 +25,15 @@ function(kp_add_executable_and_test)
+        return()
+    endif()
+
+    if(DEFINED kaeat_args_RANKS)
+        message(WARNING "Test ${kaeat_args_TARGET_NAME} will run on ${kaeat_args_RANKS} MPI ranks.")
+    endif()
+
```
Not sure we need this.
```suggestion
```

vbrunini committed:
Update tests/CMakeLists.txt

Remove debugging message.

Co-authored-by: Daniel Arndt <arndtd@ornl.gov>

vbrunini committed:
Update tests/CMakeLists.txt

Remove debugging message.

Co-authored-by: Daniel Arndt <arndtd@ornl.gov>

janciesko commented on tests/space-time-stack/test_threshold_with_imbalance.cpp:
```diff
@@ -0,0 +1,81 @@
+#include <chrono>
```
```suggestion
// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
// SPDX-FileCopyrightText: Copyright Contributors to the Kokkos project

#include <chrono>
```

janciesko commented on tests/CMakeLists.txt:
```diff
@@ -37,10 +43,18 @@ function(kp_add_executable_and_test)
+    if(DEFINED kaeat_args_RANKS)
+        add_test(
+            NAME ${kaeat_args_TARGET_NAME}
+            COMMAND mpirun -n ${kaeat_args_RANKS} $<TARGET_FILE:${kaeat_args_TARGET_NAME}>
```
```suggestion
            COMMAND ${MPIEXEC_EXECUTABLE} -n ${kaeat_args_RANKS} $<TARGET_FILE:${kaeat_args_TARGET_NAME}>
```

janciesko commented on tests/space-time-stack/test_threshold_with_imbalance.cpp:
```diff
@@ -0,0 +1,81 @@
+  std::ostringstream output;
+  std::streambuf* coutbuf = std::cout.rdbuf(output.rdbuf());
+
+  //! Run tests. @todo Replace this with Google Test.
```
```suggestion
```

janciesko commented on tests/space-time-stack/test_threshold_with_imbalance.cpp:
```diff
@@ -0,0 +1,81 @@
+};
+
+/**
+ * @test This test checks that the tool outputs deep_copy statistics.
```
```suggestion
 * @test This test checks that the tool reports imbalanced MPI regions that would be filtered out when using average time alone.
```

janciesko committed:
Update tests/space-time-stack/test_threshold_with_imbalance.cpp

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

janciesko committed:
Update tests/space-time-stack/test_threshold_with_imbalance.cpp

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

janciesko committed:
Update tests/CMakeLists.txt

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

janciesko committed:
Update tests/space-time-stack/test_threshold_with_imbalance.cpp

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

janciesko committed:
Silence -Wrestrict on deep_copy test

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>

janciesko committed:
clang-format

Signed-off-by: Jan Ciesko <jan.ciesko@gmail.com>
